src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb

```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                                                symbol As Symbol) As Boolean
             Debug.Assert(Me.DeclaringCompilation.EventQueue IsNot Nothing)
 
+            Dim change As Boolean = False
             SyncLock _diagnosticLock
-                Dim change = (variable And mask) = comparand
+                change = (variable And mask) = comparand
                 If change Then
-                    Me.DeclaringCompilation.SymbolDeclaredEvent(symbol)
-
                     If Not ThreadSafeFlagOperations.Set(variable, mask) Then
                         ' If this gets hit, then someone wrote to variable without going through this
                         ' routine, which is a bug.
                         Throw ExceptionUtilities.Unreachable
                     End If
                 End If
-                Return change
             End SyncLock
+
+            If change Then
+                ' Invoke the symbol declared event outside the lock.
+                ' Task scheduler might decide to inline the event processing, which can call back into analyzers,
```

Comment:
If anything you've found a bug in the async queue, or in the task scheduler for VS. The async queue is not supposed to allow any other code to interfere withthe code that *queues* an event (especially not code outside the compiler). Queueing an event should be a very short operation. If VS interferes with the normal task scheduling to allow a thread holding a lock to be co-opted to perform other tasks, then that is a bug in VS. It cannot be solved by a workaround here.